File: python/day5.py
import re

class Line:
    def __init__(self, spec):
        match = re.match("^([0-9]+),([0-9]+) -> ([0-9]+),([0-9]+)$", spec)
        self.x1 = int(match.group(1))
        self.y1 = int(match.group(2))
        self.x2 = int(match.group(3))
        self.y2 = int(match.group(4))

    @property
    def is_horizontal(self):
        return self.y1 == self.y2

    @property
    def is_vertical(self):
        return self.x1 == self.x2

    @property
    def xmin(self):
        return min(self.x1, self.x2)

    @property
    def xmax(self):
        return max(self.x1, self.x2)

    @property
    def ymin(self):
        return min(self.y1, self.y2)

    @property
    def ymax(self):
        return max(self.y1, self.y2)

    @property
    def xrange(self):
        return abs(self.x1 - self.x2) + 1

    @property
    def yrange(self):
        return abs(self.y1 - self.y2) + 1

    @property
    def p1(self):
        return (self.x1, self.y1)

    @property
    def p2(self):
        return (self.x2, self.y2)

    @property
    def coords(self):
        if self.is_horizontal:
            return [(self.xmin + c, self.y1) for c in range(0, self.xrange)]
        if self.is_vertical:
            return [(self.xmin, self.ymin + c) for c in range(0, self.yrange)]

        min_point = (self.xmin, self.ymin)
        positive_slope = self.p1 == min_point or self.p2 == min_point

        (y_start, y_mult) = (self.ymin, 1) if positive_slope else (self.ymax, -1)

        return [(self.xmin + c, y_start + (c * y_mult)) for c in range(0, self.xrange)]

def get_intersection_count(lines):
    all_coords = (c for l in lines for c in l.coords)
    counts = {}
    for c in all_coords:
        counts[c] = counts.get(c, 0) + 1
    intersections = [k for k, count in counts.items() if count > 1]

    return len(intersections)

def part1(input):
    lines = [Line(spec) for spec in input]
    ortho_lines = [l for l in lines if l.is_horizontal or l.is_vertical]

    return get_intersection_count(ortho_lines)

def part2(input):
    lines = [Line(spec) for spec in input]

    return get_intersection_count(lines)

if __name__ == "__main__":
    with open('day5.txt') as f:
        input = f.readlines()

    print(f'Day 5, part 1: {part1(input)}')
    print(f'Day 5, part 2: {part2(input)}')

File: typescript/day5.ts
import _ from "lodash";

type Orientation =
  | "Horizontal"
  | "Vertical";

export class Line {
  x1: number;
  x2: number;
  y1: number;
  y2: number;
  spec: string;
  constructor(spec: string) {
    const matches = (/^([0-9]+),([0-9]+) -> ([0-9]+),([0-9]+)$/).exec(spec) as RegExpMatchArray;
    this.spec = spec;

    const [x1, y1, x2, y2] = matches.slice(1).map(x => parseInt(x));

    this.x1 = x1;
    this.y1 = y1;
    this.x2 = x2;
    this.y2 = y2;

  }

  get orientation(): Orientation | null {
    if(this.x1 === this.x2) {
      return "Vertical";
    }

    if(this.y1 === this.y2) {
      return "Horizontal";
    }

    return null;
  }

  get xMax() { return Math.max(this.x1, this.x2); }
  get xMin() { return Math.min(this.x1, this.x2); }
  get yMax() { return Math.max(this.y1, this.y2); }
  get yMin() { return Math.min(this.y1, this.y2); }

  get coords(): string[] {
    if(this.orientation === "Horizontal") {
      return _.range(this.xMin, this.xMax + 1)
        .map(x => `${x},${this.y1}`);
    }
    if(this.orientation === "Vertical") {
      return _.range(this.yMin, this.yMax + 1)
        .map(y => `${this.x1},${y}`);
    }

    const range = this.xMax - this.xMin + 1;

    const [yStart, yMult] =
      (this.x1 === this.xMin && this.y1 === this.yMin)
      || (this.x2 === this.xMin && this.y2 === this.yMin)
      ? [this.yMin, 1]
      : [this.yMax, -1]

    return _.range(0, range)
      .map(offset => `${this.xMin + offset},${yStart + (offset * yMult)}`);
  }
}

function parseInput(input: string): Line[] {
  return _.chain(input.split(/\r?\n/))
    .map(x => new Line(x))
    .value();
}

function findIntersections(lines: Line[]): string[] {
  return _.chain(lines)
    .flatMap(x => x.coords)
    .groupBy(x => x)
    .pickBy((xs) => xs.length > 1)
    .keys()
    .value();
}

export function part1(input: string): number {
  const lines = parseInput(input)
    .filter(x => x.orientation !== null);

  const intersectingPoints = findIntersections(lines);

  return intersectingPoints.length;
}

export function part2(input: string): number {
  const lines = parseInput(input);

  const intersectingPoints = findIntersections(lines);

  return intersectingPoints.length;
}
